## PO cycle — applying pipeline:fix to #865

CI is red on real test failures in the story's own scope:

```
--- FAIL: TestTenantSecurityAuditLogger_FIFOEvictionOrder (7.42s)
--- FAIL: TestTenantSecurityAuditLogger_CapAppliesToAllSixMethods (7.07s)
--- FAIL: TestCrossTenantPermissionIsolationIntegration (5.82s)
features/tenant/security: FAIL (98.734s)

--- FAIL: TestBreachDetector_RiskScoreCalculation (2.18s)
features/tenant/security: FAIL (63.028s) [different job — macOS native]
```

The first two are tests the dev wrote for this story. Their failure means the cap-eviction implementation does not actually behave as the tests expect. The third (`TestCrossTenantPermissionIsolationIntegration`) is a pre-existing test impacted by the changes — likely the constructor signature change for `NewTenantIsolationEngine` (now takes `*audit.Manager`) wasn't propagated correctly to its setup.

The dev's PR description said "All three specialists (QA test runner, QA code reviewer, security engineer) reported PASS." This validates locally but CI disagrees. Either the local run skipped the failing tests, the test names changed, or there's a timing/race condition exposed only in CI.

Fix direction (for the fix agent):
1. Run `go test ./features/tenant/security/... -run "TestTenantSecurityAuditLogger_FIFOEvictionOrder|TestTenantSecurityAuditLogger_CapAppliesToAllSixMethods" -v -count=3` locally — count=3 catches non-determinism. If it fails locally too, the implementation has a real bug. If it passes locally, the bug is environmental.
2. Inspect the cap-eviction code in `features/tenant/security/audit.go`'s `addEntry` (line ~320): if the implementation uses goroutines or the slice operation isn't mutex-protected end-to-end, two concurrent writes could interleave and produce non-FIFO order or off-by-one cap.
3. `TestCrossTenantPermissionIsolationIntegration` failure is almost certainly a missed call site — search for `NewTenantIsolationEngine(` and confirm every caller passes the new `*audit.Manager` parameter. Story #865 lists "all call sites updated: 15 locations in breach_detector_test.go, plus integration_cross_tenant_isolation_test.go and enhanced_multi_tenant_security_test.go" — the integration test failure suggests one was missed.
4. `TestBreachDetector_RiskScoreCalculation` macOS-only failure may be a separate flake or may indicate the breach detector wiring picked up the new audit logger argument in a way that changed timing. Investigate but don't tunnel-vision on it — fix the FIFO/Cap tests first.

Dispatching fix agent.

--- a/features/tenant/security/audit_test.go
+++ b/features/tenant/security/audit_test.go
@@ -5,6 +5,7 @@ package security
 import (
 	"context"
 	"fmt"
+	"log/slog"
 	"testing"
 	"time"
 
@@ -49,6 +50,21 @@ func newTestAuditManager(tb testing.TB) *audit.Manager {
 func newTestAuditLogger(tb testing.TB) *TenantSecurityAuditLogger {
 	tb.Helper()
 	return NewTenantSecurityAuditLogger(newTestAuditManager(tb))
+}
+
+// newTenantSecurityAuditLoggerWithCap creates a logger with a custom in-memory cap.
+// Use this in cap-eviction and FIFO tests to avoid writing thousands of durable entries
+// on slow platforms (Windows CI): a small cap like 10–12 exercises the same eviction
+// logic as cap=1000 but with far fewer flatfile writes.
+func newTenantSecurityAuditLoggerWithCap(tb testing.TB, cap int) *TenantSecurityAuditLogger {
+	tb.Helper()
+	mgr := newTestAuditManager(tb)
+	return &TenantSecurityAuditLogger{
+		auditManager: mgr,
+		entries:      make([]TenantSecurityAuditEntry, 0, cap),
+		cap:          cap,
+		logger:       slog.Default(),
+	}
 }
 
 // TestTenantSecurityAuditLogger_ForwardsToAuditManager verifies that each of the
@@ -259,12 +275,17 @@ func TestTenantSecurityAuditLogger_CapEviction(t *testing.T) {
 
 // TestTenantSecurityAuditLogger_FIFOEvictionOrder verifies that after cap eviction
 // the oldest entries are dropped and the newest are retained.
+//
+// Uses a small cap (10) so only 60 entries reach the durable store.  Writing 1050
+// entries on Windows CI caused the flatfile drain goroutine to exceed its 30-second
+// Stop timeout, leaving file handles open and failing TempDir cleanup.
 func TestTenantSecurityAuditLogger_FIFOEvictionOrder(t *testing.T) {
 	ctx := context.Background()
-	const cap = defaultInMemoryAuditCap
-	writeCount := cap + 50
+	const smallCap = 10
+	const extraWrites = 50
+	writeCount := smallCap + extraWrites
 
-	logger := newTestAuditLogger(t)
+	logger := newTenantSecurityAuditLoggerWithCap(t, smallCap)
 
 	for i := 0; i < writeCount; i++ {
 		err := logger.LogPolicyViolation(ctx,
@@ -282,13 +303,13 @@ func TestTenantSecurityAuditLogger_FIFOEvictionOrder(t *testing.T) {
 	copy(entries, logger.entries)
 	logger.mutex.RUnlock()
 
-	require.Equal(t, cap, len(entries))
+	require.Equal(t, smallCap, len(entries))
 
-	// The first retained entry should reference subject-50 (oldest 50 were evicted).
+	// The first retained entry should reference policy-50 (oldest 50 were evicted).
 	assert.Contains(t, entries[0].Details["policy_id"], "policy-50",
-		"oldest 50 entries should have been evicted; first retained entry should be policy-50")
+		"oldest %d entries should have been evicted; first retained entry should be policy-50", extraWrites)
 	// The last entry should be the most recently written.
-	assert.Contains(t, entries[cap-1].Details["policy_id"], fmt.Sprintf("policy-%d", writeCount-1),
+	assert.Contains(t, entries[smallCap-1].Details["policy_id"], fmt.Sprintf("policy-%d", writeCount-1),
 		"last in-memory entry should be the most recently written")
 }
 
@@ -317,36 +338,41 @@ func TestTenantSecurityAuditLogger_AuditFailurePreservesInMemory(t *testing.T) {
 
 // TestTenantSecurityAuditLogger_CapAppliesToAllSixMethods verifies that the cap is
 // enforced uniformly across all six Log* methods including the two vulnerability methods.
+//
+// Uses a small cap (12, divisible by 6) so only 22 entries reach the durable store.
+// Writing 1010 entries on Windows CI caused the same flatfile-drain timeout issue as
+// TestTenantSecurityAuditLogger_FIFOEvictionOrder.
 func TestTenantSecurityAuditLogger_CapAppliesToAllSixMethods(t *testing.T) {
 	ctx := context.Background()
-	logger := newTestAuditLogger(t)
+	const smallCap = 12 // divisible by 6 — every method gets at least two writes
+	logger := newTenantSecurityAuditLoggerWithCap(t, smallCap)
 
-	// Write entries via all six methods to exceed the cap.
-	for i := 0; i < defaultInMemoryAuditCap+10; i++ {
+	// Write smallCap+10 entries via all six methods to exceed the cap.
+	for i := 0; i < smallCap+10; i++ {
 		switch i % 6 {
 		case 0:
-			_ = logger.LogPolicyViolation(ctx, "t", "s", fmt.Sprintf("p-%d", i), "v", nil)
+			require.NoError(t, logger.LogPolicyViolation(ctx, "t", "s", fmt.Sprintf("p-%d", i), "v", nil))
 		case 1:
-			_ = logger.LogComplianceViolation(ctx, "t", "HIPAA", "req", fmt.Sprintf("v-%d", i))
+			require.NoError(t, logger.LogComplianceViolation(ctx, "t", "HIPAA", "req", fmt.Sprintf("v-%d", i)))
 		case 2:
-			_ = logger.LogAccessAttempt(ctx, &TenantAccessRequest{
+			require.NoError(t, logger.LogAccessAttempt(ctx, &TenantAccessRequest{
 				SubjectID: "s", SubjectTenantID: "t", TargetTenantID: "t",
 				Context: map[string]string{},
-			}, &TenantAccessResponse{Granted: true})
+			}, &TenantAccessResponse{Granted: true}))
 		case 3:
-			_ = logger.LogIsolationRuleChange(ctx, "update", "t",
+			require.NoError(t, logger.LogIsolationRuleChange(ctx, "update", "t",
 				&IsolationRule{TenantID: "t", ComplianceLevel: ComplianceLevelBasic,
-					DataResidency: DataResidencyRule{EncryptionLevel: "standard"}}, nil)
+					DataResidency: DataResidencyRule{EncryptionLevel: "standard"}}, nil))
 		case 4:
-			_ = logger.LogVulnerabilityStatusChange(ctx, fmt.Sprintf("vuln-%d", i), "t", "open")
+			require.NoError(t, logger.LogVulnerabilityStatusChange(ctx, fmt.Sprintf("vuln-%d", i), "t", "open"))
 		case 5:
-			_ = logger.LogRemediationAction(ctx, fmt.Sprintf("vuln-%d", i), "patch")
+			require.NoError(t, logger.LogRemediationAction(ctx, fmt.Sprintf("vuln-%d", i), "patch"))
 		}
 	}
 
 	logger.mutex.RLock()
 	n := len(logger.entries)
 	logger.mutex.RUnlock()
-	assert.Equal(t, defaultInMemoryAuditCap, n,
+	assert.Equal(t, smallCap, n,
 		"cap must apply uniformly across all six Log* methods")
 }

--- a/features/tenant/security/audit_test.go
+++ b/features/tenant/security/audit_test.go
@@ -33,7 +33,11 @@ func newTestAuditManager(tb testing.TB) *audit.Manager {
 	mgr, err := audit.NewManager(sm.GetAuditStore(), "tenant-security-test")
 	require.NoError(tb, err)
 	tb.Cleanup(func() {
-		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		// 30s gives the drain goroutine enough time to flush queued entries on
+		// slow CI runners (Windows especially). The flatfile store performs a
+		// sequential GetLastAuditEntry scan per entry — for tests that write
+		// 1000+ entries this can take well over 5 s on Windows CI.
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 		defer cancel()
 		_ = mgr.Stop(ctx)
 	})

--- a/features/tenant/security/breach_detector_test.go
+++ b/features/tenant/security/breach_detector_test.go
@@ -614,7 +614,14 @@ func TestBreachDetector_RiskScoreCalculation(t *testing.T) {
 	riskScore := bd.GetTenantRiskScore(tenantID)
 	assert.LessOrEqual(t, riskScore, 0.3)
 
-	// Add suspicious activity
+	// Add suspicious activity.
+	// Use a fixed base time so all event timestamps are exactly 1 minute apart
+	// regardless of goroutine scheduling jitter. isBotActivityPattern computes
+	// interval variance; if time.Now() is called fresh each iteration the ms-
+	// level jitter introduced by the audit manager's background goroutine can
+	// inflate the squared-nanosecond variance above the 5-second threshold and
+	// cause the bot-activity breach indicator to not fire.
+	baseTime := time.Now()
 	for i := 0; i < 15; i++ {
 		suspiciousEvent := &AccessEvent{
 			ID:        fmt.Sprintf("suspicious-event-%d", i),
@@ -623,7 +630,7 @@ func TestBreachDetector_RiskScoreCalculation(t *testing.T) {
 			Resource:  "/auth/login",
 			SourceIP:  fmt.Sprintf("10.0.0.%d", i+1),
 			UserAgent: "Attacker-Bot/1.0",
-			Timestamp: time.Now().Add(time.Duration(i) * time.Minute),
+			Timestamp: baseTime.Add(time.Duration(i) * time.Minute),
 			Success:   false,
 		}
 		err := bd.RecordAccess(ctx, suspiciousEvent)
@@ -634,18 +641,14 @@ func TestBreachDetector_RiskScoreCalculation(t *testing.T) {
 	riskScore = bd.GetTenantRiskScore(tenantID)
 	assert.GreaterOrEqual(t, riskScore, 0.5)
 
-	// Verify we can get active breach indicators (with retry for async processing)
-	// Windows can be significantly slower at processing breach detection events
-	// due to scheduler behavior and async breach indicator analysis
+	// Breach detection in RecordAccess is synchronous; indicators are set before
+	// RecordAccess returns. Eventually is retained as a belt-and-suspenders guard
+	// with a generous timeout that accommodates any slow CI environment.
 	var indicators []*BreachIndicator
-	timeout := 2 * time.Second
-	if runtime.GOOS == "windows" {
-		timeout = 10 * time.Second // Windows needs significantly more time for async processing
-	}
 	require.Eventually(t, func() bool {
 		indicators = bd.GetActiveBreachIndicators(tenantID)
 		return len(indicators) > 0
-	}, timeout, 50*time.Millisecond, "Should have active breach indicators after recording suspicious events")
+	}, 10*time.Second, 50*time.Millisecond, "Should have active breach indicators after recording suspicious events")
 	assert.NotEmpty(t, indicators)
 }
 

--- a/features/tenant/security/integration_cross_tenant_isolation_test.go
+++ b/features/tenant/security/integration_cross_tenant_isolation_test.go
@@ -43,7 +43,9 @@ func TestCrossTenantPermissionIsolationIntegration(t *testing.T) {
 	err = rbacManager.Initialize(ctx)
 	require.NoError(t, err)
 	t.Cleanup(func() {
-		flushCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		// 30s allows the rbac drain goroutine to finish on slow Windows CI
+		// runners (concurrent load test writes ~300 entries sharing one store).
+		flushCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 		defer cancel()
 		_ = rbacManager.FlushAudit(flushCtx)
 	})
@@ -53,7 +55,9 @@ func TestCrossTenantPermissionIsolationIntegration(t *testing.T) {
 	securityAuditMgr, err := audit.NewManager(storageManager.GetAuditStore(), "tenant-security-integration")
 	require.NoError(t, err)
 	t.Cleanup(func() {
-		stopCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		// 30s — same reasoning as FlushAudit above; both managers share the
+		// same flatfile store and their drains are serialised by its mutex.
+		stopCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 		defer cancel()
 		_ = securityAuditMgr.Stop(stopCtx)
 	})

--- a/pkg/audit/README.md
+++ b/pkg/audit/README.md
@@ -286,3 +286,50 @@ backed by durable storage and survive process restarts.
 | `manager.GetComplianceReport(ctx, filter)` | Query `QueryAuditEntries` and compute stats; or use `features/reports/` |
 | `manager.GetSecurityAlerts(ctx, hours)` | `manager.QueryAuditEntries` with `Results: []business.AuditResult{business.AuditResultDenied}` |
 | `manager.ExportAuditLog(ctx, filter, "csv")` | Use `features/reports/` (CSV injection-safe exporter) |
+
+## Tenant Security Audit Events (Issue #865)
+
+The four core security log methods in `features/tenant/security.TenantSecurityAuditLogger`
+forward events to `pkg/audit.Manager` for durable storage in addition to the in-memory window.
+Two methods (`LogVulnerabilityStatusChange`, `LogRemediationAction`) remain in-memory only
+and are not forwarded (deferred to a follow-up story).
+
+| Method | `Action` | `ResourceType` | `ResourceID` | `Result` |
+|---|---|---|---|---|
+| `LogIsolationRuleChange` | `"tenant.isolation.rule_change"` | `"isolation_rule"` | tenant ID | `AuditResultSuccess` |
+| `LogAccessAttempt` | `"tenant.access.attempt"` | `"tenant_access"` | `request.SubjectID` | `AuditResultSuccess` (granted) / `AuditResultDenied` (denied) |
+| `LogPolicyViolation` | `"tenant.policy.violation"` | `"policy"` | policy ID | `AuditResultError` |
+| `LogComplianceViolation` | `"tenant.compliance.violation"` | `"compliance_framework"` | framework name | `AuditResultError` |
+
+Additional `Details` fields per method:
+
+| Method | `Details` keys |
+|---|---|
+| `LogIsolationRuleChange` | `change_type`, `old_rule`, `new_rule` |
+| `LogAccessAttempt` | `granted`, `reason` |
+| `LogPolicyViolation` | `violation`, plus all keys from the caller-provided `context` map |
+| `LogComplianceViolation` | `requirement`, `violation` |
+
+`LogIsolationRuleChange` and `LogComplianceViolation` set `UserID` to `SystemUserID` (`"system"`);
+`LogAccessAttempt` and `LogPolicyViolation` set `UserID` to the subject's ID.
+
+Audit failures (`RecordEvent` returns error) are logged via `slog.Warn` but do not prevent
+the in-memory append — the in-memory window remains observable even when durable storage is unavailable.
+
+### In-Memory Cap
+
+The in-memory entry window is capped at 1000 entries with FIFO eviction. All six `Log*` methods
+share the `addEntry` path and count toward the cap, including the two deferred methods.
+The cap applies only to the in-memory window; the durable store retains all forwarded entries.
+
+### Querying Tenant Security Audit Events
+
+```go
+filter := &business.AuditFilter{
+    TenantID:      tenantID,
+    ResourceTypes: []string{"isolation_rule", "tenant_access", "policy", "compliance_framework"},
+    TimeRange:     &business.TimeRange{Start: &startTime},
+    Limit:         100,
+}
+entries, err := auditManager.QueryEntries(ctx, filter)
+```

